src/Hint/NewType.hs

```diff
@@ -65,3 +77,45 @@ hasNoStrategy _                      = False
 withNewtype :: Deriving a -> Deriving a
 withNewtype (Deriving l Nothing rs)  = Deriving l (Just $ DerivNewtype l) rs
 withNewtype d                        = d
+
+data WarnNewtype = WarnNewtype
+    { newDecl :: Hs.LHsDecl Hs.GhcPs
+    , insideType :: Hs.HsType Hs.GhcPs
+    }
+
+-- | Given a declaration, returns the suggested \"newtype\"ized declaration following these guidelines:
+-- * @MagicHash@'d stuff is __ignored__ - @data X = X Int#@
+-- * @ExistentialQuantification@ stuff is __ignored__ - @data X = forall t. X t@
+-- * Constructors with (nonempty) constraints are __ignored__ - @data X a = (Eq a) => X a@
+-- * Single field constructors get newtyped - @data X = X Int@ -> @newtype X = X Int@
+-- * Single record field constructors get newtyped - @data X = X {getX :: Int}@ -> @newtype X = X {getX :: Int}@
+-- * All other declarations are ignored.
+singleSimpleFieldNew :: Hs.LHsDecl Hs.GhcPs -> Maybe WarnNewtype
+singleSimpleFieldNew (Hs.L loc (Hs.TyClD ext decl@(Hs.DataDecl _ name _ _ def@(Hs.HsDataDefn _ Hs.DataType ctx _ _ [Hs.L _ constructor] derives))))
+    | Just inType <- simpleCons constructor = 
+        Just WarnNewtype 
+              { newDecl = (Hs.L loc $ Hs.TyClD ext decl {Hs.tcdDataDefn = def {Hs.dd_ND = Hs.NewType}})
+              , insideType = inType
+              }
+singleSimpleFieldNew _ = Nothing
+
+-- TODO: check for MAGIC#HASH
+-- TODO: get tests to pass
+--
+-- | Checks whether its argument is a \"simple constructor\" (see criteria in 'singleSimpleFieldNew')
+-- returning the type inside the constructor if it is. This is needed for bang/MagicHash analysis.
+simpleCons :: Hs.ConDecl Hs.GhcPs -> Maybe (Hs.HsType Hs.GhcPs)
+simpleCons (Hs.ConDeclH98 _ _ _ [] context (Hs.PrefixCon [Hs.L _ inType]) _) 
+    | emptyOrNoContext context = Just inType
+simpleCons (Hs.ConDeclH98 _ _ _ [] context (Hs.RecCon (Hs.L _ [Hs.L _ (Hs.ConDeclField _ [_] (Hs.L _ inType) _)])) _)
+    | emptyOrNoContext context = Just inType
+simpleCons _ = Nothing
+
+warnBang :: Hs.HsType Hs.GhcPs -> Bool
```

Comment:
TODO: @googleson78 This is redundant! It can easily be replaced by http://hackage.haskell.org/package/ghc-lib-parser-8.8.0.20190424/docs/HsTypes.html#v:getBangStrictness

Reply:
This is not redundant, we would still need to do pattern matching and such.